src/utils.rs

```diff
@@ -0,0 +1,114 @@
+use blake2::digest::{Update, VariableOutput};
+use blake2::Blake2bVar;
+use std::cmp::min;
+
+// Oracles
+
+/// Takes as input a hash and inverse probability $np$ and returns a
+/// uniformly distributed integer in [1, np] (c.f. Appendix B, Alba paper).
+pub fn oracle_uniform(hash: &[u8], n: usize) -> usize {
+    // Computes the integer reprensation of hash* modulo n when n is not a
+    // power of two. *(up to 8 bytes, in little endian)
+    fn mod_non_power_of_2(hash: &[u8], n: usize) -> usize {
+        fn log_base2(x: usize) -> usize {
+            usize::BITS as usize - x.leading_zeros() as usize - 1
+        }
+        let epsilon_fail: usize = 1 << 40; // roughly 1 trillion
+        let k: usize = log_base2(n * epsilon_fail);
+        let k_prime: usize = 1 << k;
+        let d: usize = k_prime.div_ceil(n);
+
+        let i = mod_power_of_2(hash, k_prime);
+
+        if i >= d * n {
+            panic!("failed: i = {}, d = {}, n = {}, k = {}", i, d, n, k);
```

Comment:
I think we should return an `Option` from this function instead of panicking, similarly to the prove() function.

Reply:
`Option` or `Result<T, E>`?

Reply:
Just `Option` I think. There is only one reason to fail, and this error will not propagated to the user anyway since the DFS needs to continue.

Reply:
We could also reference section B in the paper that talks about it.